hal/src/thumbv7em/icm.rs

```diff
@@ -0,0 +1,1612 @@
+    #[inline]
+    fn icm(&self) -> &RegisterBlock {
+        &self.icm
+    }
+
+    /// Configuration
+    #[inline]
+    fn cfg(&self) -> &CFG {
+        &self.icm().cfg
+    }
+
+    /// Control
+    #[inline]
+    fn ctrl(&self) -> &CTRL {
+        &self.icm().ctrl
+    }
+
+    /// Region Descriptor Area Start Address
+    #[inline]
+    fn dscr(&self) -> &DSCR {
+        &self.icm().dscr
+    }
+
+    /// Region Hash Area Start Address
+    #[inline]
+    fn hash(&self) -> &HASH {
+        &self.icm().hash
+    }
+
+    /// Interrupt Disable
+    #[inline]
+    fn idr(&self) -> &IDR {
+        &self.icm().idr
+    }
+
+    /// Interrupt Enable
+    #[inline]
+    fn ier(&self) -> &IER {
+        &self.icm().ier
+    }
+
+    /// Interrupt Mask
+    #[inline]
+    pub fn imr(&self) -> &IMR {
+        &self.icm().imr
+    }
+
+    /// Interrupt Status
+    #[inline]
+    fn isr(&self) -> &ISR {
+        &self.icm().isr
+    }
+
+    /// Status
+    #[inline]
+    fn sr(&self) -> &SR {
+        &self.icm().sr
+    }
+
+    /// Undefined Access Status
+    #[inline]
+    fn uasr(&self) -> &UASR {
+        &self.icm().uasr
+    }
+}
+
+/// Struct useful for returning the interrupt status
+/// of the ICM. Provides methods for easy parsing of
+/// all the regions or via the `bitmask` argument
+/// narrow it down to the specific set of [`IcmRegionNum`]
+/// of interest.
+pub struct IcmInterrupt {
+    interrupt_vector: u32,
+}
+
+impl IcmInterrupt {
+    /// Region Status Updated interrupt status
+    #[inline]
+    pub fn get_rsu_int(&self, bitmask: u8) -> u8 {
+        (self.interrupt_vector >> 20 & 0x0f) as u8 & bitmask
+    }
+
+    /// Region End bit Condition Detected interrupt status
+    #[inline]
+    pub fn get_rec_int(&self, bitmask: u8) -> u8 {
+        (self.interrupt_vector >> 16 & 0x0f) as u8 & bitmask
+    }
+
+    /// Region Wrap Condition detected interrupt status
+    #[inline]
+    pub fn get_rwc_int(&self, bitmask: u8) -> u8 {
+        (self.interrupt_vector >> 12 & 0x0f) as u8 & bitmask
+    }
+
+    /// Region Bus Error interrupt status
+    #[inline]
+    pub fn get_rbe_int(&self, bitmask: u8) -> u8 {
+        (self.interrupt_vector >> 8 & 0x0f) as u8 & bitmask
+    }
+
+    /// Region Digest Mis interrupt status
+    #[inline]
+    pub fn get_rdm_int(&self, bitmask: u8) -> u8 {
+        (self.interrupt_vector >> 4 & 0x0f) as u8 & bitmask
+    }
+
+    /// Region Hash Completed interrupt status
+    #[inline]
+    pub fn get_rhc_int(&self, bitmask: u8) -> u8 {
+        (self.interrupt_vector & 0x0f) as u8 & bitmask
+    }
+}
+impl Default for IcmInterrupt {
+    fn default() -> Self {
+        Self {
+            interrupt_vector: 0,
+        }
+    }
+}
+
+/// Struct useful for returning the interrupt status
+/// of the ICM. Provides methods for easy parsing of
+/// the region specific [`IcmRegionNum`]
+pub struct IcmRegionInterrupt<I: IcmRegionNum> {
+    region: PhantomData<I>,
+    interrupt_vector: u32,
+}
```

Comment:
The type parameter makes it more difficult, but I think this also might be another candidate for a `bitflags` type, or something similar.

Reply:
Not as straight-forward, but combining the previous style and bitflags became much nicer than it was